### PR TITLE
Correct the default collation of Babelfish data types except TEXT and NTEXT

### PR DESCRIPTION
**Pull request by Deepesh125**
### Description

Correct the default collation of Babelfish data types.

This commit contains following changes:
1. Move collation creation script from babelfishpg_tsql extension to babelfishpg_common extension.
2. Implements hook define_type_default_collation_hook to override the default collation of Babelfish data types being defined.
3. Move collation related GUC, babelfishpg_tsql.server_collation_name and babelfishpg_tsql.default_locale to babelfishpg_common. Initialisation of these GUCs required to load collations info correctly which would not happen if they are part of babelfishpg_tsql because we load babelfishpg_common before babelfishpg_tsql.
4. Changes in sys objects to avoid collation conflict error or to have correct collation
5. Introduced new function sys.babelfish_update_collation_to_default(schema_name varchar, table_name varchar, column_name varchar) in upgrade script, babelfishpg_tsql--2.2.0--2.3.0.sql, which could be used to set the collation of any column of any view to server's default collation.
6. Introduced new function sys.babelfish_update_collation_to_default(typname) in upgrade script babelfishpg_common--2.2.0--2.3.0.sql which could be used to update the default collation of Babelfish data types.
7. Introduced two new GUCs namely babelfishpg_tsql.restored_server_collation_name and babelfishpg_tsql.restored_default_locale to restore the user defined value of babelfishpg_tsql.server_collation_name and babelfishpg_tsql.default_locale respectively during major version upgrade.
8. Upgrade script, babelfishpg_tsql--2.2.0--2.3.0.sql, Fixed default collation of all sys views/catalog using above procedure

This commit also enables upgrade path 14.3/14.5 -> 14.6 -> 15.1 and 14.6 -> 15.1. 

### Testing

- Use case based: Added few new tests cases
- Boundary conditions: Not required
- Arbitrary inputs: Not required
- Negative test scenarios: Not required
- Minor version upgrade tests: Relying on github actions 
- Major version upgrade tests: Relying on github actions. 
- Performance tests: Have not seen big impact on sla test cases except **few tests failure which will be fixed**
- Tooling impact: **To be done**
- Client tests

### MVU Testing

MVU works fine. But just need to ensure that we persist customer defined GUCs like `babelfishpg_tsql.server_collation_name` and `babelfishpg_tsql.default_locale` GUC during upgrade.

Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/70
 
### Issues Resolved

BABEL-3315

Authored-by: Dipesh Dhameliya <dddhamel@amazon.com>
Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/configuration/upgrade-test-configuration.yml [hunk 1]
```diff
@@ -7,4 +49,18 @@ upgrade-version: [[
...
+    upgrade-type: null
+  },
+  { 
+    version: 14.6,
+    upgrade-type: minor 
+  },
+  { 
+    version: target.latest,
+    upgrade-type: major 
+  }
```
Comment (rishabhtanwar29) [inline .github/configuration/upgrade-test-configuration.yml:65]: Let's also add direct upgrade path from 14.6 to latest

File: .github/composite-actions/build-modified-postgres/action.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ inputs:
   engine_branch:
     description: 'Engine Branch'
     required: no
-    default: 'latest'
+    default: 'jira-babel-3801'
```
Comment (rishabhtanwar29) [inline .github/composite-actions/build-modified-postgres/action.yml]: We don't need these temp changes as branches name matches b/w engine and extension.

File: .github/scripts/clone_engine_repo [hunk 3]
```diff
@@ -78,6 +78,10 @@ else
             BRANCH_OPTION=''
         fi
     fi
+    if [ "$BRANCH" = "jira-babel-3801" ]; then
```
Comment (rishabhtanwar29) [inline .github/scripts/clone_engine_repo]: Same here, we don't need this temp change.

File: contrib/babelfishpg_common/src/collation.c [hunk 4]
```diff
@@ -1405,4 +1412,74 @@ get_collation_callbacks(void)
...
+{
+	PG_RETURN_OID(get_server_collation_oid_internal(false));
+}
+
+PG_FUNCTION_INFO_V1(babelfish_update_server_collation_name);
+
+/*
+ * babelfish_update_server_collation_name - corresponding to sys.babelfish_update_server_collation_name() function
+ * which would be available and strictly be used during 1.x to 2.3 upgrade.
+ */
```
Comment (Deepesh125) [inline contrib/babelfishpg_common/src/collation.c]: @shah-nirmit we also need to update this comment.

File: contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--2.3.0--3.0.0.sql [hunk 5]
```diff
@@ -0,0 +1,31 @@
+-- complain if script is sourced in psql, rather than via ALTER EXTENSION
+\echo Use "ALTER EXTENSION ""babelfishpg_common"" UPDATE TO '3.0.0'" to load this file. \quit
+
+SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false);
+
+/* This helper function would only be useful and strictly be used during 1.x to 2.3 upgrade. */
```
Comment (Deepesh125) [inline contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--2.3.0--3.0.0.sql]: This also needs correction.

File: .github/configuration/upgrade-test-configuration.yml (hunk 1 above)
Comment (Deepesh125) [reply .github/configuration/upgrade-test-configuration.yml:65]: Done

File: .github/scripts/clone_engine_repo (hunk 3 above)
Comment (Deepesh125) [reply .github/scripts/clone_engine_repo]: Done!

File: .github/composite-actions/build-modified-postgres/action.yml (hunk 2 above)
Comment (Deepesh125) [reply .github/composite-actions/build-modified-postgres/action.yml]: Done!

**Review by rishabhtanwar29 — DISMISSED** [review]

**Review by surendravishnoi — APPROVED** [review]

**Review by rishabhtanwar29 — APPROVED** [review]

**Pull request merged**